### PR TITLE
[Uptime] Feature/synthetics integration package

### PR DESCRIPTION
**Pull request by dominiqueclarke**
Please label this PR with one of the following labels, depending on the scope of your change:
- Enhancement

## What does this PR do?

This PR adds the spec for the Elastic Synthetics integration package, which includes support for HTTP, ICMP, and TCP monitors.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/master/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [ ] I have verified that all data streams collect metrics or logs.  (Collects synthetics)
- [x] I have added an entry to my package's `changelog.yml` file.

## Author's Checklist

- [ ] Ensure that field data types are compatible with heartbeat config

## How to test this PR locally

1. Install Elastic Package (https://github.com/elastic/elastic-package). 
2. Navigate to `packages/synthetics` in the integration repo
3. Run `elastic-package build` to build the synthetics package
4. Run `elastic-package stack up --version 8.0.0-SNAPSHOT` to serve kibana, es, and package registry
5. Log into kibana and navigate to Management -> Fleet
6. Once in fleet, navigate to the Integrations tab and search for Synthetics
7. Click Add Integration once on the Synthetics page to test. You should see the default form builder UI with all our configuration. 

## Related issues
Relates elastic/kibana#91062

### CODE REVIEW

**Review by dominiqueclarke — COMMENTED** [review]
PR is work in progress. TCP and ICMP fields will need to be adjusted to only include relevant fields. Right now everything is based off HTTP, so HTTP is the only spec that is truly in a shipable state.

File: packages/synthetics/changelog.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# newer versions go on top
+- version: "0.0.1" # unreleased
```
Comment (dominiqueclarke) [inline packages/synthetics/changelog.yml:2]: Starting from 0.0.1, in case we need to do additional patches before releasing the first public version as `0.1.0`.

File: packages/synthetics/data_stream/tcp/manifest.yml [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        required: false
+        show_user: true
+      - name: check.send
+        type: text
+        title: Request payload
+        multi: false
+        required: false
+        show_user: true
+      - name: check.receive
+        type: yaml
```
Comment (dominiqueclarke) [inline packages/synthetics/data_stream/tcp/manifest.yml:83]: Should the type for check.receive be yaml or text @andrewvc ?

File: packages/synthetics/changelog.yml (hunk 1 above)
Comment (dominiqueclarke) [reply packages/synthetics/changelog.yml:2]: I've also see other teams append `-dev` on the end of a release when they don't intend to publish it.

File: packages/synthetics/data_stream/http/fields/jolokia-autodiscover.yml [hunk 3]
```diff
@@ -0,0 +1,35 @@
+- name: jolokia.agent.version
```
Comment (shahzad31) [inline packages/synthetics/data_stream/http/fields/jolokia-autodiscover.yml:1]: is this file needed? name seems weird.

File: packages/synthetics/manifest.yml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        title: HTTP
+        description: Perform an HTTP check
+      - type: synthetics/tcp
+        title: TCP
+        description: Perform a TCP check
+      - type: synthetics/icmp
+        title: ICMP
+        description: Perform an ICMP check
+conditions:
+  kibana.version: "^7.10.0"
```
Comment (shahzad31) [inline packages/synthetics/manifest.yml:25]: this seems awkward, any reason why this is 7.10.0?

File: packages/synthetics/data_stream/http/fields/base-fields.yml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+- name: name
```
Comment (dominiqueclarke) [inline packages/synthetics/data_stream/http/fields/base-fields.yml:1]: All files in `/data_stream/*/fields` are mappings. They are not auto generated, but were copied over from mappings in heartbeat.

File: packages/synthetics/data_stream/http/fields/ecs.yml [hunk 6]
```diff
@@ -0,0 +1,1371 @@
+- name: "@timestamp"
```
Comment (dominiqueclarke) [inline packages/synthetics/data_stream/http/fields/ecs.yml:1]: I actually forget the strategy we used for getting the ecs fields since @andrewvc and Justin took care of this before I took over the project. (I was on the zoom but pretty sure most of it was going over my head at that point). Andrew can you remind me how we gained these fields? Did we just look at the mapping in the heartbeat index in dev tools?

File: packages/synthetics/data_stream/http/fields/http.yml [hunk 7]
```diff
@@ -0,0 +1,91 @@
+- name: http
```
Comment (dominiqueclarke) [inline packages/synthetics/data_stream/http/fields/http.yml:1]: I'd say it's most important to review the heartbeat specific mappings, including `http.yml`, `tcp.yml`, and `socks5.yml`, `tls.yml`, and `icmp.yml`.

File: packages/synthetics/data_stream/icmp/agent/stream/icmp.yml.hbs [hunk 8]
```diff
@@ -0,0 +1,10 @@
+type: {{type}}
```
Comment (dominiqueclarke) [inline packages/synthetics/data_stream/icmp/agent/stream/icmp.yml.hbs:1]: Determines how the user rendered fields map to the policy.

File: packages/synthetics/data_stream/http/agent/stream/http.yml.hbs [hunk 9]
```diff
@@ -0,0 +1,18 @@
+type: {{type}}
```
Comment (dominiqueclarke) [inline packages/synthetics/data_stream/http/agent/stream/http.yml.hbs:1]: Determines how the user rendered fields map to the policy.

File: packages/synthetics/data_stream/http/manifest.yml [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+title: synthetic monitor check
+release: experimental
+dataset: http
+streams:
+  - input: synthetics/http
+    title: Synthetic monitor check
+    description: Monitor the health of an HTTP endpoint
+    template_path: http.yml.hbs
+    enabled: false
+    vars:
```
Comment (dominiqueclarke) [inline packages/synthetics/data_stream/http/manifest.yml:11]: Determines fields rendered to the user when using the form builder. For the custom UI, this now becomes the interface we are interacting with to pass the data. We are provided with a mutable object that contains keys mapping to each one of the `name` fields within this list. The data we pass in is then parsed as the provided `type` and included in the policy.

File: packages/synthetics/data_stream/icmp/manifest.yml [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+title: synthetic monitor check
+release: experimental
+dataset: icmp
+streams:
+  - input: synthetics/icmp
+    title: Synthetic monitor check
+    description: Perform an ICMP check
+    template_path: icmp.yml.hbs
+    enabled: false
+    vars:
```
Comment (dominiqueclarke) [inline packages/synthetics/data_stream/icmp/manifest.yml:11]: Determines fields rendered to the user when using the form builder. For the custom UI, this now becomes the interface we are interacting with to pass the data. We are provided with a mutable object that contains keys mapping to each one of the `name` fields within this list. The data we pass in is then parsed as the provided `type` and included in the policy.

File: packages/synthetics/data_stream/tcp/manifest.yml [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+title: synthetic monitor check
+release: experimental
+dataset: tcp
+streams:
+  - input: synthetics/tcp
+    title: Synthetic monitor check
+    description: Monitor the health of an TCP endpoint
+    template_path: tcp.yml.hbs
+    enabled: false
+    vars:
```
Comment (dominiqueclarke) [inline packages/synthetics/data_stream/tcp/manifest.yml:11]: Determines fields rendered to the user when using the form builder. For the custom UI, this now becomes the interface we are interacting with to pass the data. We are provided with a mutable object that contains keys mapping to each one of the `name` fields within this list. The data we pass in is then parsed as the provided `type` and included in the policy.

File: packages/synthetics/manifest.yml [hunk 13]
```diff
@@ -0,0 +1,31 @@
+format_version: 1.0.0
+name: synthetics
+title: Elastic Synthetics
+description: Monitor services availability.
+version: 0.0.1
```
Comment (dominiqueclarke) [inline packages/synthetics/manifest.yml:5]: Starting from 0.0.1, in case we need to do additional patches before releasing the first public version as 0.1.0. I've also seen other teams append `-dev` for versions they don't intend to release.

File: packages/synthetics/manifest.yml [hunk 14]
```diff
@@ -0,0 +1,31 @@
+format_version: 1.0.0
```
Comment (dominiqueclarke) [inline packages/synthetics/manifest.yml:1]: I actually don't know what this means in relation to the version @ycombinator

File: packages/synthetics/data_stream/tcp/agent/stream/tcp.yml.hbs [hunk 15]
```diff
@@ -0,0 +1,12 @@
+type: {{type}}
```
Comment (dominiqueclarke) [inline packages/synthetics/data_stream/tcp/agent/stream/tcp.yml.hbs:1]: Determines how the user rendered fields map to the policy.

File: packages/synthetics/manifest.yml (hunk 14 above)
Comment (ycombinator) [reply packages/synthetics/manifest.yml:1]: This field refers to the version of the package spec. It basically says that this package has a structure conformant with version 1.0.0 of the package spec. At the moment 1.0.0 is the only acceptable value. If/when the package spec gets a version bump this field will start becoming relevant. If you want even more details, see https://github.com/elastic/package-spec#specification-versioning.

File: packages/synthetics/data_stream/http/fields/jolokia-autodiscover.yml (hunk 3 above)
Comment (dominiqueclarke) [reply packages/synthetics/data_stream/http/fields/jolokia-autodiscover.yml:1]: jolokia mappings are included in the heartbeat index mappings.

File: packages/synthetics/manifest.yml (hunk 4 above)
Comment (dominiqueclarke) [reply packages/synthetics/manifest.yml]: Updated.

**Review by shahzad31 — APPROVED** [review]
Fields and mappings LGTM !!

File: packages/synthetics/data_stream/http/fields/beat.yml [hunk 16]
```diff
@@ -0,0 +1,9 @@
+- name: fields
+  type: object
+  object_type: keyword
+  description: >
+    Contains user configurable fields.
+
+- name: timeseries.instance
```
Comment (andrewvc) [inline packages/synthetics/data_stream/http/fields/beat.yml:7]: I don't think we use this  field. let's remove it. CC @shahzad31 for confirmation

File: packages/synthetics/data_stream/icmp/fields/beat.yml [hunk 17]
```diff
@@ -0,0 +1,9 @@
+- name: fields
+  type: object
+  object_type: keyword
+  description: >
+    Contains user configurable fields.
+
+- name: timeseries.instance
```
Comment (andrewvc) [inline packages/synthetics/data_stream/icmp/fields/beat.yml:7]: I don't think we use this  field. let's remove it. CC @shahzad31 for confirmation

File: packages/synthetics/data_stream/icmp/fields/base-fields.yml [hunk 18]
```diff
@@ -0,0 +1,9 @@
+- name: name
+  type: constant_keyword
```
Comment (andrewvc) [inline packages/synthetics/data_stream/icmp/fields/base-fields.yml:2]: Same concern here, I don't think any of these fields are used. Are we certain they came from heartbeat? I don't see them in the the `fields.yml` in `heartbeat` in the beats repo.

File: packages/synthetics/data_stream/http/fields/common.yml [hunk 19]
```diff
@@ -0,0 +1,58 @@
+- name: monitor
+  type: group
+  description: >
+    Common monitor fields.
+
+  fields:
+    - name: type
+      type: constant_keyword
```
Comment (andrewvc) [inline packages/synthetics/data_stream/http/fields/common.yml:8]: This is correct usage of `constant_keyword`

File: packages/synthetics/data_stream/http/fields/ecs.yml (hunk 6 above)
Comment (andrewvc) [reply packages/synthetics/data_stream/http/fields/ecs.yml:1]: We manually edited the list of ECS fields down to ones that would potentially make sense in heartbeat. So, we're good here :)

File: packages/synthetics/data_stream/http/fields/jolokia-autodiscover.yml (hunk 3 above)
Comment (andrewvc) [reply packages/synthetics/data_stream/http/fields/jolokia-autodiscover.yml:1]: Jolokia is an autodiscover provider that we support. I don't know if anyone's using it, but we do support it since it comes free for all beats. This would include jolokia metadata.

File: packages/synthetics/data_stream/icmp/fields/common.yml [hunk 20]
```diff
@@ -0,0 +1,58 @@
+- name: monitor
+  type: group
+  description: >
+    Common monitor fields.
+
+  fields:
+    - name: type
+      type: constant_keyword
+      value: http
```
Comment (andrewvc) [inline packages/synthetics/data_stream/icmp/fields/common.yml:9]: The `value` here should be `icmp`, since all ICMP monitors will have a type field value of `icmp`

File: packages/synthetics/data_stream/icmp/manifest.yml (hunk 11 above)
Comment (andrewvc) [reply packages/synthetics/data_stream/icmp/manifest.yml:11]: I see, so, this just becomes an empty nested object that we populate via our custom UI. Is that right?

File: packages/synthetics/data_stream/tcp/fields/base-fields.yml [hunk 21]
```diff
@@ -0,0 +1,9 @@
+- name: name
+  type: constant_keyword
+  description: Dataset name.
```
Comment (andrewvc) [inline packages/synthetics/data_stream/tcp/fields/base-fields.yml:3]: I don't think we need this file either

File: packages/synthetics/docs/README.md [hunk 22]
```diff
@@ -0,0 +1,4 @@
+# Elastic Synthetics Integration
+
+This integration sets up required assets for Synthetics to use central management.
```
Comment (andrewvc) [inline packages/synthetics/docs/README.md:3]: We should improve this README at some point, but I think we can do that in a subsequent PR.

For now can we just clean it up a bit to: "Run lightweight synthetic checks on a schedule to ensure your HTTP, TCP, and ICMP services are up. Pairs with the Uptime Kibana UI."

**Review by andrewvc — COMMENTED** [review]
Generally looks good, though a few fields seem to have been added that needn't have been.

File: packages/synthetics/data_stream/http/fields/beat.yml (hunk 16 above)
Comment (shahzad31) [reply packages/synthetics/data_stream/http/fields/beat.yml:7]: ++

File: packages/synthetics/data_stream/icmp/fields/beat.yml (hunk 17 above)
Comment (shahzad31) [reply packages/synthetics/data_stream/icmp/fields/beat.yml:7]: ++

File: packages/synthetics/data_stream/icmp/fields/base-fields.yml (hunk 18 above)
Comment (shahzad31) [reply packages/synthetics/data_stream/icmp/fields/base-fields.yml:2]: ++

File: packages/synthetics/data_stream/http/fields/base-fields.yml (hunk 5 above)
Comment (dominiqueclarke) [reply packages/synthetics/data_stream/http/fields/base-fields.yml:1]: I agree I believe it's here mistakenly.

File: packages/synthetics/changelog.yml (hunk 1 above)
Comment (ycombinator) [reply packages/synthetics/changelog.yml:2]: I'd suggest keeping it simple going with @dominiqueclarke's original suggestion in this thread. That is, start with `0.0.1` in this PR. Then can bump it up to `0.0.2`, `0.0.3`, etc. in subsequent PRs as you develop the package. 

As you merge PRs with these versions into the `integrations` repo, these package versions will automatically get published to the `snapshot` branch of the package registry, accessible at http://epr-snapshot.elastic.co/. This can be used to share the package across multiple developers for testing, etc. by pointing your Kibana instance to this registry URL.

Then, when you are happy enough with the package for the first public version, bump up the version to `0.1.0`.

Note that if you are not interested in sharing the package via the `snapshot` registry, you can make PRs to the integrations repo without bumping up the version. This means you accumulate multiple changes in the same version, say `0.0.1`. In this case, to test the package, developers would need to checkout the `integrations` repo and run `elastic-package build` from the `packages/synthetics` folder and bounce the local stack to pick up the latest package changes.

File: packages/synthetics/data_stream/http/agent/stream/http.yml.hbs [hunk 23]
```diff
@@ -0,0 +1,23 @@
...
+check.request.headers: {{check.request.headers}}
+check.request.body: {{check.request.body}}
+check.response.status: {{check.response.status}}
+check.response.headers: {{check.response.headers}}
+check.response.body.negative: {{check.response.body.negative}}
+check.response.body.positive: {{check.response.body.positive}}
+ssl.certificate: {{ssl.certificate}}
+ssl.certificate_authorities: {{ssl.certificate_authorities}}
+ssl.key: {{ssl.key}}
+ssl.key_passphrase: {{ssl.key_passphrase}}
```
Comment (dominiqueclarke) [inline packages/synthetics/data_stream/http/agent/stream/http.yml.hbs:22]: @andrewvc , remind me, did we decide to keep key_passphrase or no?

Comment (andrewvc) [reply packages/synthetics/data_stream/http/agent/stream/http.yml.hbs:22]: Yes, let's keep it

File: packages/synthetics/data_stream/http/agent/stream/http.yml.hbs [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+proxy_url: {{proxy_url}}
+tags: {{tags}}
+response.include_headers: {{response.include_headers}}
+response.include_body: {{response.include_body}}
+check.request.method: {{check.request.method}}
+check.request.headers: {{check.request.headers}}
+check.request.body: {{check.request.body}}
+check.response.status: {{check.response.status}}
+check.response.headers: {{check.response.headers}}
+check.response.body.negative: {{check.response.body.negative}}
```
Comment (andrewvc) [inline packages/synthetics/data_stream/http/agent/stream/http.yml.hbs:17]: Let's wait to merge till we can see if we can get a conditional in here for this. If the answer is no, then let's make the change in the beat.

**Comment by andrewvc** [comment]
Generally looks good, I'd prefer to test this with the UI, which requires the conditionals for HTTP positive/negative matches however
